ecosystem/indexer-grpc/indexer-grpc-utils/src/in_memory_cache.rs

```diff
@@ -0,0 +1,439 @@
+// Copyright © Aptos Foundation
+
+use crate::compression_util::{CacheEntry, StorageFormat};
+use anyhow::Context;
+use aptos_protos::transaction::v1::Transaction;
+use dashmap::DashMap;
+use itertools::Itertools;
+use prost::Message;
+use redis::AsyncCommands;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+// Internal lookup retry interval for in-memory cache.
+const IN_MEMORY_CACHE_LOOKUP_RETRY_INTERVAL_MS: u64 = 10;
+// Max cache size in bytes: 5 GB.
+const MAX_IN_MEMORY_CACHE_CAPACITY_IN_BYTES: u64 = 5_000_000_000;
+// Max cache entry TTL: 30 seconds.
+// const MAX_IN_MEMORY_CACHE_ENTRY_TTL: u64 = 30;
+// Warm-up cache entries. Pre-fetch the cache entries to warm up the cache.
+const WARM_UP_CACHE_ENTRIES: u64 = 20_000;
+const MAX_REDIS_FETCH_BATCH_SIZE: usize = 500;
+
+#[derive(Debug, Clone, Copy)]
+struct CacheMetadata {
+    total_size_in_bytes: u64,
+    latest_version: u64,
+    first_version: u64,
+}
+
+/// InMemoryCache is a simple in-memory cache that stores the protobuf Transaction.
+pub struct InMemoryCache {
+    /// Cache maps the cache key to the deserialized Transaction.
+    cache: Arc<DashMap<u64, Arc<Transaction>>>,
```

Comment:
do you want to have some batching here to avoid memory fragmentation?

Reply:
in that case, something like `Arc<RwLock<BTreeMap<u64, Vec<Transaction>>>>` and we batch transactions in each `V` right?

Reply:
That's what I'm thinking.. 
or having a deque to store the data and having a BTreeMap to store the index, 
or still using the DashMap, each item is X transactions, and the last `latest_version % X` transactions stored separately.

Not sure if any of them is good idea though.

Reply:
That sounds good! Spent some time to think about it and the challenge is like what you said, what would be the most performant implementation for us. 

considering we have multiple use cases like this, e.g., event stream requires in-memory cache too. I'll do a follow-up PR with some benches + tests so that we can find the best impl for us and expose this as a generic library.

cc @just-in-chang